File: algorithms/053.maximum-subarray.cpp
class Solution {
public:
    // DP solution : O(n)
    int maxSubArray(vector<int>& nums) {
        if(nums.size() == 0)
            return 0;
            
        int max_val = nums[0], DP;
        for (int i=0; i < nums.size(); i++){
            DP = (i == 0) ? nums[i] : max(DP + nums[i], nums[i]);
            max_val = max(DP, max_val);
        }
        return max_val;
    }
};

File: algorithms/053.maximum-subarray.java
public class Solution {
	public int maxSubArray(int[] nums) {
		if (nums.length == 0)
			return 0;
		
		int maxVal = nums[0];
		int DP = nums[0];
		
		for (int i = 1; i < nums.length; i++) {
			DP = Math.max(DP + nums[i], nums[i]);
			maxVal = Math.max(maxVal, DP);
		}
		
		return maxVal;
	}
}
